hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/StreamCommitWatcher.java

```diff
@@ -140,11 +162,55 @@ public XceiverClientReply streamWatchForCommit(long commitIndex)
     }
   }
 
+  void releaseBuffersOnException() {
+    adjustBuffers(xceiverClient.getReplicatedMinCommitIndex());
+  }
+
+  private void adjustBuffers(long commitIndex) {
+    List<Long> keyList = commitIndexSet.keySet().stream()
+        .filter(p -> p <= commitIndex).collect(Collectors.toList());
+    if (!keyList.isEmpty()) {
+      releaseBuffers(keyList);
+    }
+  }
+
+  private long releaseBuffers(List<Long> indexes) {
+    Preconditions.checkArgument(!commitIndexSet.isEmpty());
+    for (long index : indexes) {
+      Preconditions.checkState(commitIndexSet.containsKey(index));
+      final List<ByteBuffer> buffers
+          = commitIndexSet.remove(index);
+      long length =
+          buffers.stream().mapToLong(buf -> (buf.limit() - buf.position()))
+              .sum();
+      totalAckDataLength += length;
+      // clear the future object from the future Map
+      final CompletableFuture<ContainerCommandResponseProto> remove =
+          futureMap.remove(totalAckDataLength);
+      if (remove == null) {
+        LOG.error("Couldn't find required future for " + totalAckDataLength);
+        for (Long key : futureMap.keySet()) {
+          LOG.error("Existing acknowledged data: " + key);
+        }
+      }
+      Preconditions.checkNotNull(remove);
+      for (ByteBuffer byteBuffer : buffers) {
+        bufferPool.remove(byteBuffer);
```

Comment:
This is very expensive!  It will compare every byte inside the buffer.  Anyway, we should not have a bufferPool here.